File: cloud/oauth-server/src/main/java/pers/acp/admin/oauth/domain/ApplicationDomain.java
package pers.acp.admin.oauth.domain;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import pers.acp.admin.oauth.base.BaseDomain;
import pers.acp.admin.oauth.entity.Application;
import pers.acp.admin.oauth.po.ApplicationPO;
import pers.acp.admin.oauth.repo.ApplicationRepository;
import pers.acp.admin.oauth.repo.UserRepository;
import pers.acp.core.CommonTools;
import pers.acp.springboot.core.exceptions.ServerException;

import javax.persistence.criteria.Predicate;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

/**
 * @author zhang by 13/01/2019
 * @since JDK 11
 */
@Service
public class ApplicationDomain extends BaseDomain {

    private final ApplicationRepository applicationRepository;

    @Autowired
    public ApplicationDomain(UserRepository userRepository, ApplicationRepository applicationRepository) {
        super(userRepository);
        this.applicationRepository = applicationRepository;
    }

    @Transactional
    public Application doCreate(ApplicationPO applicationPO) {
        Application application = new Application();
        application.setAppname(applicationPO.getAppname());
        application.setSecret(CommonTools.getUuid());
        application.setAccessTokenValiditySeconds(applicationPO.getAccessTokenValiditySeconds());
        application.setRefreshTokenValiditySeconds(applicationPO.getRefreshTokenValiditySeconds());
        application.setCovert(true);
        return applicationRepository.save(application);
    }

    @Transactional
    public Application doUpdate(ApplicationPO applicationPO) throws ServerException {
        Optional<Application> applicaitonOptional = applicationRepository.findById(applicationPO.getId());
        if (applicaitonOptional.isEmpty()) {
            throw new ServerException("找不到信息");
        }
        Application application = applicaitonOptional.get();
        application.setAppname(applicationPO.getAppname());
        application.setAccessTokenValiditySeconds(applicationPO.getAccessTokenValiditySeconds());
        application.setRefreshTokenValiditySeconds(applicationPO.getRefreshTokenValiditySeconds());
        return applicationRepository.save(application);
    }

    @Transactional
    public Application doUpdateSecret(String appId) throws ServerException {
        Optional<Application> applicaitonOptional = applicationRepository.findById(appId);
        if (applicaitonOptional.isEmpty()) {
            throw new ServerException("找不到信息");
        }
        Application application = applicaitonOptional.get();
        application.setSecret(CommonTools.getUuid());
        return applicationRepository.save(application);
    }

    @Transactional
    public void doDelete(List<String> idList) {
        applicationRepository.deleteByIdInAndCovert(idList, true);
    }

    public Page<Application> doQuery(ApplicationPO applicationPO) {
        return applicationRepository.findAll((Specification<Application>) (root, query, criteriaBuilder) -> {
            List<Predicate> predicateList = new ArrayList<>();
            if (!CommonTools.isNullStr(applicationPO.getAppname())) {
                predicateList.add(criteriaBuilder.like(root.get("appname").as(String.class), "%" + applicationPO.getAppname() + "%"));
            }
            return criteriaBuilder.and(predicateList.toArray(new Predicate[]{}));
        }, buildPageRequest(applicationPO.getQueryParam()));
    }

}

File: cloud/oauth-server/src/main/kotlin/pers/acp/admin/oauth/domain/ApplicationDomain.kt
package pers.acp.admin.oauth.domain

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.Page
import org.springframework.security.oauth2.provider.OAuth2Authentication
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import pers.acp.admin.oauth.base.OauthBaseDomain
import pers.acp.admin.oauth.entity.Application
import pers.acp.admin.oauth.po.ApplicationPo
import pers.acp.admin.oauth.po.ApplicationQueryPo
import pers.acp.admin.oauth.repo.ApplicationRepository
import pers.acp.admin.oauth.repo.UserRepository
import pers.acp.core.CommonTools
import pers.acp.spring.boot.exceptions.ServerException

import javax.persistence.criteria.Predicate

/**
 * @author zhang by 13/01/2019
 * @since JDK 11
 */
@Service
@Transactional(readOnly = true)
class ApplicationDomain @Autowired
constructor(userRepository: UserRepository, private val applicationRepository: ApplicationRepository) : OauthBaseDomain(userRepository) {

    fun getAppList(): MutableList<Application> = applicationRepository.findAllByOrderByIdentifyAscAppNameAsc()

    fun getOwnAppList(user: OAuth2Authentication): MutableList<Application> {
        val currUserInfo = findCurrUserInfo(user.name) ?: throw ServerException("无法获取当前用户信息")
        return if (isSuper(currUserInfo)) {
            applicationRepository.findAllByOrderByIdentifyAscAppNameAsc()
        } else {
            applicationRepository.findById(user.oAuth2Request.clientId).let {
                if (it.isPresent) {
                    mutableListOf(it.get())
                } else {
                    mutableListOf()
                }
            }
        }
    }

    @Transactional
    fun doCreate(applicationPo: ApplicationPo): Application =
            Application(
                    appName = applicationPo.appName!!,
                    secret = CommonTools.getUuid32(),
                    scope = applicationPo.scope?.trim()?.replace("，", ",") ?: "",
                    identify = applicationPo.identify?.trim() ?: "",
                    accessTokenValiditySeconds = applicationPo.accessTokenValiditySeconds,
                    refreshTokenValiditySeconds = applicationPo.refreshTokenValiditySeconds,
                    covert = true
            ).let {
                applicationRepository.save(it)
            }

    @Transactional
    @Throws(ServerException::class)
    fun doUpdate(applicationPo: ApplicationPo): Application =
            applicationRepository.getOne(applicationPo.id!!).copy(
                    appName = applicationPo.appName!!,
                    scope = applicationPo.scope?.trim()?.replace("，", ",") ?: "",
                    identify = applicationPo.identify?.trim() ?: "",
                    accessTokenValiditySeconds = applicationPo.accessTokenValiditySeconds,
                    refreshTokenValiditySeconds = applicationPo.refreshTokenValiditySeconds
            ).let {
                applicationRepository.save(it)
            }

    @Transactional
    @Throws(ServerException::class)
    fun doUpdateSecret(appId: String): Application =
            applicationRepository.getOne(appId).copy(
                    secret = CommonTools.getUuid32()
            ).let {
                applicationRepository.save(it)
            }

    @Transactional
    fun doDelete(idList: MutableList<String>) = applicationRepository.deleteByIdInAndCovert(idList, true)

    fun doQuery(applicationQueryPo: ApplicationQueryPo): Page<Application> =
            applicationRepository.findAll({ root, _, criteriaBuilder ->
                val predicateList: MutableList<Predicate> = mutableListOf()
                if (!CommonTools.isNullStr(applicationQueryPo.appName)) {
                    predicateList.add(criteriaBuilder.like(root.get<Any>("appName").`as`(String::class.java), "%" + applicationQueryPo.appName + "%"))
                }
                criteriaBuilder.and(*predicateList.toTypedArray())
            }, buildPageRequest(applicationQueryPo.queryParam!!))

    fun getApp(appId: String): Application? = applicationRepository.findById(appId).orElse(null)

}
